[feature] Implement Iperf check #385 (#394)
openwisp/openwisp-monitoring: README.rst

.. code-block:: python

    from django.conf import settings

>     from celery.schedules import crontab

Author: thanks! I will update :+1:

    # here we show how to configure celery with redis but you can
    # use other brokers if you want, consult the celery docs
    # Celery TIME_ZONE should be equal to django TIME_ZONE
    # In order to schedule run_iperf_checks on the correct time intervals
    CELERY_TIMEZONE = TIME_ZONE
    CELERY_BROKER_URL = 'redis://localhost/1'
    CELERY_BEAT_SCHEDULE = {
        # Celery beat configuration for auto checks ie ping & config applied
        'run_checks': {
            'task': 'openwisp_monitoring.check.tasks.run_checks',
            'schedule': timedelta(minutes=5),
        },
        # Delete old WifiSession
        'delete_wifi_clients_and_sessions': {
            'task': 'openwisp_monitoring.monitoring.tasks.delete_wifi_clients_and_sessions',

>             'schedule': timedelta(days=180),

Author: By mistake 😅 I'll add :+1:

    # here we show how to configure celery with redis but you can
    # use other brokers if you want, consult the celery docs
    # Celery TIME_ZONE should be equal to django TIME_ZONE
    # In order to schedule run_iperf_checks on the correct time intervals
    CELERY_TIMEZONE = TIME_ZONE
    CELERY_BROKER_URL = 'redis://localhost/1'
    CELERY_BEAT_SCHEDULE = {
       'run_checks': {

>         'task': 'openwisp_monitoring.check.tasks.run_checks',

Reviewer: the indentation is not right here

Author: Done 👍

        # Executes only ping & config check every 5 min
        'schedule': timedelta(minutes=5),
        'args': (
            [  # Checks path
                'openwisp_monitoring.check.classes.Ping',
                'openwisp_monitoring.check.classes.ConfigApplied',
            ],
        ),

>         'relative': True,

Reviewer: aren't we missing the instructions to run the iperf check?

Author: I think it is mention [`here`](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#3-configure-iperf-settings)

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/access-technology.png
  :align: center

Iperf
~~~~~

> +--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

Reviewer 2: @Aryamanz29 don't make these tables too wide. As you can observe here it is difficult to read. Please break the content into multiple lines.

| **measurement**:   | ``iperf``                                                                                                                                                                                                                      |
+--------------------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+

> | **types**:         | ``int`` (iperf_result, retransmits, total_packets, lost_packets), ``float`` (sent_bps_tcp, received_bps_tcp, sent_bytes_tcp, received_bytes_tcp, sent_bps_udp, sent_bytes_udp, jitter, lost_percent)                           |

Reviewer 2: Especially here, I would break it into two more more lines, one for `int` and other for `float`.

Author: I'll update :+1:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/access-technology.png
  :align: center

Iperf
~~~~~

> +--------------------+----------------------------------------------------------------------------------------------------------------------------------------+

Author: After adding `|` to my table, it looks like this : 

![Screenshot from 2022-07-08 01-23-35](https://user-images.githubusercontent.com/56113566/177860603-acd2bc75-cd62-4ad4-b338-a208073e2b76.png)

| **fields**:        | | ``iperf_result``, ``sent_bps_tcp``, ``received_bps_tcp``, ``sent_bytes_tcp``, ``received_bytes_tcp``, ``retransmits``,  |
|                    | | ``sent_bps_udp``, ``sent_bytes_udp``, ``jitter``, ``total_packets``, ``lost_packets``, ``lost_percent``                 |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+
| **configuration**: | ``iperf``                                                                                                                 |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+
| **charts**:        | | ``bandwidth_tcp``, ``transfer_tcp``, ``retransmits``, ``bandwidth_udp``                                                 |
|                    | | ``transfer_udp``, ``jitter``, ``datagram``, ``datagram_loss``                                                           |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+

**Bandwidth (TCP)**:

> .. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/bandwidth-tcp.png

Author: Sure :+1:

|                    | | ``sent_bps_udp``, ``sent_bytes_udp``, ``jitter``, ``total_packets``, ``lost_packets``, ``lost_percent``                 |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+
| **configuration**: | ``iperf``                                                                                                                 |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+
| **charts**:        | | ``bandwidth_tcp``, ``transfer_tcp``, ``retransmits``, ``bandwidth_udp``                                                 |
|                    | | ``transfer_udp``, ``jitter``, ``datagram``, ``datagram_loss``                                                           |
+--------------------+---------------------------------------------------------------------------------------------------------------------------+

**Bandwidth (TCP)**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/bandwidth-tcp.png
  :align: center

**Transfer (TCP)**:

> .. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/transfer-tcp.png

Reviewer: let's do those changes to the charts that we discussed (have bandwidth and transfer as simple lines with no fill and have 2 lines in each chart, one for TCP and one for UDP, then update the README here and screenshots.

**Bandwidth (TCP)**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/bandwidth-tcp.png
  :align: center

**Transfer (TCP)**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/transfer-tcp.png
  :align: center

**Retransmits**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/retransmits.png
  :align: center

**Bandwidth (UDP)**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/bandwidth-udp.png
  :align: center

**Transfer (UDP)**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/transfer-udp.png
  :align: center

**Jitter**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/jitter.png
  :align: center

**Datagram**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/datagram.png
  :align: center

**Datagram loss**:

.. figure:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/datagram-loss.png
  :align: center

**Note:** Iperf charts uses ``connect_points=True`` in 

> `default chart configuration <#openwisp_monitoring_charts>`_ that joins it's individual chart data points.

Author: I and @nemesisdesign had meet in which he suggested me to add this about `iperf charts`.

> `default chart configuration <#openwisp_monitoring_charts>`_ that joins it's individual chart data points.

Reviewer 2: Okay, thank you!

quickly to events happening in the network and informs the user promptly
if there's anything that is not working as intended.

Iperf
~~~~~

> This check provides network performance measurements such as maximum achievable bandwidth, jitter, datagram loss etc of the device using `iperf3 utility <https://iperf.fr/>`_. 

Author: Got it :+1:

It also supports tuning of various parameters related to timing, buffers and protocols (TCP, UDP with IPv4 and IPv6).

This check is ``disabled`` by **default**, but you may choose to enable auto creation of this check by setting
`OPENWISP_MONITORING_AUTO_IPERF <#OPENWISP_MONITORING_AUTO_IPERF>`_ to ``True``.

You can also change the default values used for iperf checks using
`OPENWISP_MONITORING_IPERF_CHECK_CONFIG <#OPENWISP_MONITORING_IPERF_CHECK_CONFIG>`_ setting.

> Instructions to configure Iperf Check

Reviewer 2: I would create a heading `Usage Instructions` and move this section under the sub-heading `How to configure iperf check`.  This is make it similar to REAMDE of openwisp-controller. 

Wait for @nemesisdesign opinion before going ahead with this change.

Reviewer: I am not sure I understood correctly but if you look at the result generated at https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#iperf-1, it doesn't look good (this heading is bigger than the one above), so anything better than that will be ok for me, I am pretty sure we have to give it a last round of refinement anyway @pandafy.

Author: Improved docs :+1:

-------------------------------------

1. Register your device to OpenWISP
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Register your device to OpenWISP and make sure `iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ is installed on the device if not run :

.. code-block:: shell

    opkg install iperf3

2. Enable secure SSH access from OpenWISP to your devices
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> We have already well covered this in `How to configure push updates section of openwisp-controller <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_.

Reviewer 2: Always use relative  hyperlinks 

```suggestion
 Follow the steps in `"How to configure push updates" section of  the openwisp-controller documentation <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_ to allow SSH access to 
 you device from OpenWISP.
```

> **Note:** Make sure device connection is enabled & working with right update strategy ie. ``OpenWISP SSH``.

Reviewer 2: Isn't it **OpenWRT SSH** ?

```suggestion
**Note:** Make sure device connection is enabled & working with right update strategy i.e. ``OpenWRT SSH``.
```

Author: Yes :+1:

.. image:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/enable-openwisp-ssh.png
  :alt: Enable ssh access from openwisp to device
  :align: center
  
3. Configure Iperf settings
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Configure iperf servers in `openwisp settings <https://github.com/openwisp/openwisp-monitoring/blob/master/tests/openwisp2/settings.py>`_ , The host can be specified by hostname, IPv4 literal, or IPv6 literal.

> For eg.

Reviewer 2: ```suggestion
For example:
```

.. code-block:: python

    OPENWISP_MONITORING_IPERF_SERVERS = {
    # Public iperf servers also available: https://iperf.fr/iperf-servers.php#public-servers
    # '<org-pk>': ['<ORG_IPERF_SERVER>']
    'a9734710-db30-46b0-a2fc-01f01046fe4f': ['speedtest.uztelecom.uz'],
    'z9734710-db30-46b0-a2fc-01f01046fe4f': ['192.168.5.109'],
    'c9734710-db30-46b0-a2fc-01f01046fe4f': ['2001:db8::1'],
    }

> 4. Run the check

Reviewer 2: We should also point it out here to configure the celery beat.

Author: Noted

quickly to events happening in the network and informs the user promptly
if there's anything that is not working as intended.

Iperf
~~~~~

This check provides network performance measurements such as maximum achievable bandwidth, 
jitter, datagram loss etc of the device using `iperf3 utility <https://iperf.fr/>`_. 

This check is **disabled by default**. You can enable auto creation of this check by setting the
`OPENWISP_MONITORING_AUTO_IPERF <#OPENWISP_MONITORING_AUTO_IPERF>`_ to ``True``.

It also supports tuning of various parameters.

You can also change the parameters used for iperf checks (e.g. timing, port, buffer, etc) using the
`OPENWISP_MONITORING_IPERF_CHECK_CONFIG <#OPENWISP_MONITORING_IPERF_CHECK_CONFIG>`_ setting.

Usage Instructions
------------------

How to configure iperf check
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Register your device to OpenWISP
###################################

> Register your device to OpenWISP and make sure `iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ is installed on the device if not run :

Author: 👍

.. code-block:: python

   OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {
       'a9734710-db30-46b0-a2fc-01f01046fe4f': {
           'host': ['iperf1.openwisp.io', 'iperf2.openwisp.io', '192.168.5.2'],
           # All three parameters (username, password, rsa_publc_key)
           # are required for iperf authentication
           'username': 'iperfuser',
           'password': 'iperfpass',
           # Add RSA public key without any headers
           # ie. -----BEGIN PUBLIC KEY-----, -----BEGIN END KEY-----
           'rsa_public_key': (
               """
               MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwuEm+iYrfSWJOupy6X3N
               dxZvUCxvmoL3uoGAs0O0Y32unUQrwcTIxudy38JSuCccD+k2Rf8S4WuZSiTxaoea
               6Du99YQGVZeY67uJ21SWFqWU+w6ONUj3TrNNWoICN7BXGLE2BbSBz9YaXefE3aqw
               GhEjQz364Itwm425vHn2MntSp0weWb4hUCjQUyyooRXPrFUGBOuY+VvAvMyAG4Uk
               msapnWnBSxXt7Tbb++A5XbOMdM2mwNYDEtkD5ksC/x3EVBrI9FvENsH9+u/8J9Mf
               2oPl4MnlCMY86MQypkeUn7eVWfDnseNky7TyC0/IgCXve/iaydCCFdkjyo1MTAA4
               BQIDAQAB
               """
           ),
           # relative to /root directory

>            'rsa_public_key_path' : '/tmp/iperf-public-key.pem',

Author: This is the location where the `client (ie. openwrt device)` stores it's `rsa_public_key` for `iperf check`. By default, it is `/tmp/iperf-rsa-public.pem` and this path is required for iperf3 auth, but we can also change this using [OPENWISP_IPERF_CHECK_CONFIG](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#openwisp_monitoring_iperf_check_config). More info : https://github.com/openwisp/openwisp-monitoring/issues/414#issuecomment-1189408176

>            'rsa_public_key_path' : '/tmp/iperf-public-key.pem',

Author: Iperf command basically doing  `f'echo "{rsa_public_key}" > {rsa_public_key_path}` and after completion of check removing key from the device `rm {rsa_public_key_path}`

>            'rsa_public_key_path' : '/tmp/iperf-public-key.pem',

Reviewer: `/tmp/iperf-public-key.pem` does not need to be configurable at this level, there's no usefulness in this, just set `/tmp/iperf-public-key.pem` to a variable and use the variable throughout the rest of the implementation.

>            'rsa_public_key_path' : '/tmp/iperf-public-key.pem',

Author: Ok :+1:

.. code-block:: python

    CELERY_BEAT_SCHEDULE = {
        # Other celery beat configurations

        # Celery beat configuration for iperf check
        'run_iperf_checks': {
            'task': 'openwisp_monitoring.check.tasks.run_checks',
            # https://docs.celeryq.dev/en/latest/userguide/periodic-tasks.html#crontab-schedules
            # Executes check every 5 mins from 00:00 AM to 6:00 AM (night)
            'schedule': crontab(minute='*/5', hour='0-6'),
            # Iperf check path
            'args': (['openwisp_monitoring.check.classes.Iperf'],),
            'relative': True,
        }
    }

**Note:** We recommended to configure this check for night or during non peak traffic times to not interfere with standard traffic.

4. Run the check
################

This should happen automatically if you have celery running in the background. For testing, you can
run this check manually using the `run_checks <#run_checks>`_ command. After that, you should see the
iperf network measurements charts.

.. image:: https://github.com/openwisp/openwisp-monitoring/raw/docs/docs/1.1/iperf-charts.png
  :alt: Iperf network measurement charts

> Configure Iperf check for authentication

Reviewer: ```suggestion
Iperf authentication
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> By default iperf check runs without any kind of **authentication**, But we can configure it to use **RSA authentication** 
> between the **client** and the **server** to restrict the connections to the server & only allow legitimate clients.

Reviewer: ```suggestion
By default iperf check runs without any kind of **authentication**,
in this section we will explain how to configure **RSA authentication** 
between the **client** and the **server** to restrict connections
to authenticated clients.
```

At Iperf server
###############

1. Generate RSA keypair
^^^^^^^^^^^^^^^^^^^^^^^

.. code-block:: shell

   openssl genrsa -des3 -out private.pem 2048
   openssl rsa -in private.pem -outform PEM -pubout -out public.pem
   openssl rsa -in private.pem -out private_not_protected.pem -outform PEM

> After running above mention commands, the public key will be contained in the 

Reviewer: ```suggestion
After running the commands mentioned above, the public key will be stored in ``public.pem`` 
```

Keep it short, apply a similar change also to ``private_not_protected.pem``, I would also rename that as `private_key.pem`, while I would rename `public.pem` to `public_key.pem`.

Author: Sure :)

At client (openwrt device)
##########################

1. Install iperf3-ssl
^^^^^^^^^^^^^^^^^^^^^

Install `iperf3-ssl openwrt package <https://openwrt.org/packages/pkgdata/iperf3-ssl>`_ instead of normal
`iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ which comes without any authentication.

You may also check your installed **iperf3 openwrt package** features:

.. code-block:: shell

   root@vm-openwrt:~ iperf3 -v
   iperf 3.7 (cJSON 1.5.2)
   Linux vm-openwrt 4.14.171 #0 SMP Thu Feb 27 21:05:12 2020 x86_64
   Optional features available: CPU affinity setting, IPv6 flow label, TCP congestion algorithm setting, 
   sendfile / zerocopy, socket pacing, authentication # contains 'authentication'

2. Configure iperf check auth parameters
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Now, add the following iperf authentication parameters 
to `OPENWISP_MONITORING_IPERF_CHECK_CONFIG <#OPENWISP_MONITORING_IPERF_CHECK_CONFIG>`_ 
in `openwisp settings <https://github.com/openwisp/openwisp-monitoring/blob/master/tests/openwisp2/settings.py>`_

.. code-block:: python

    OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {

>     # All three parameters are required

Reviewer: indentation is wrong here

    'username': {'default': 'iperfuser'},

>     'password': {'default': 'iperfpass'},

Reviewer: what is `default`?

Author: I made it to look consistent with [`ping config override setting`](https://github.com/openwisp/openwisp-monitoring#openwisp_monitoring_ping_check_config), Now I have changed `iperf config setting` according to https://github.com/openwisp/openwisp-monitoring/issues/418

    # RSA public key without any headers
    # ie. -----BEGIN PUBLIC KEY-----, -----BEGIN END KEY-----
    'rsa_public_key': {
        'default': """
        MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwuEm+iYrfSWJOupy6X3N 
        dxZvUCxvmoL3uoGAs0O0Y32unUQrwcTIxudy38JSuCccD+k2Rf8S4WuZSiTxaoea 
        6Du99YQGVZeY67uJ21SWFqWU+w6ONUj3TrNNWoICN7BXGLE2BbSBz9YaXefE3aqw 
        GhEjQz364Itwm425vHn2MntSp0weWb4hUCjQUyyooRXPrFUGBOuY+VvAvMyAG4Uk 
        msapnWnBSxXt7Tbb++A5XbOMdM2mwNYDEtkD5ksC/x3EVBrI9FvENsH9+u/8J9Mf 
        2oPl4MnlCMY86MQypkeUn7eVWfDnseNky7TyC0/IgCXve/iaydCCFdkjyo1MTAA4 
        BQIDAQAB
        """
    },

>    }

Reviewer: We already have a setting to configure iperf, why are you adding a new one?

Can we compress everything in one setting?

Eg:

```python
OPENWISP_MONITORING_IPERF = {
    'a9734710-db30-46b0-a2fc-01f01046fe4f': [
        {
            'host': ['iperf1.openwisp.io', 'iperf2.openwisp.io'],
            'username': 'iperfuser',
            'password': 'iperfpassword',
            'rsa_public_key': (
"""
MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAwuEm+iYrfSWJOupy6X3N 
dxZvUCxvmoL3uoGAs0O0Y32unUQrwcTIxudy38JSuCccD+k2Rf8S4WuZSiTxaoea 
6Du99YQGVZeY67uJ21SWFqWU+w6ONUj3TrNNWoICN7BXGLE2BbSBz9YaXefE3aqw 
GhEjQz364Itwm425vHn2MntSp0weWb4hUCjQUyyooRXPrFUGBOuY+VvAvMyAG4Uk 
msapnWnBSxXt7Tbb++A5XbOMdM2mwNYDEtkD5ksC/x3EVBrI9FvENsH9+u/8J9Mf 
2oPl4MnlCMY86MQypkeUn7eVWfDnseNky7TyC0/IgCXve/iaydCCFdkjyo1MTAA4 
BQIDAQAB
""")
            'client_options': {
                'port': '6209',
                'udp': {
                    'bitrate': '20M'
                },
                'tcp': {
                    'bitrate': '0'
                }
            }
        }
    ],
}
``` 

Does it make sense?

Author: Yes, it would be better to have a single setting for `iperf check` :+1:

jitter, datagram loss etc of the device using `iperf3 utility <https://iperf.fr/>`_. 

This check is **disabled by default**. You can enable auto creation of this check by setting the
`OPENWISP_MONITORING_AUTO_IPERF <#OPENWISP_MONITORING_AUTO_IPERF>`_ to ``True``.

It also supports tuning of various parameters.

You can also change the parameters used for iperf checks (e.g. timing, port, username, password, rsa_publc_key etc) 
using the `OPENWISP_MONITORING_IPERF_CHECK_CONFIG <#OPENWISP_MONITORING_IPERF_CHECK_CONFIG>`_ setting.

Usage Instructions
------------------

How to configure iperf check
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

1. Register your device to OpenWISP
###################################

Register your device to OpenWISP and make sure `iperf3 openwrt package 
<https://openwrt.org/packages/pkgdata/iperf3>`_ is installed on the device if not run :

.. code-block:: shell

    opkg install iperf3

2. Enable secure SSH access from OpenWISP to your devices
#########################################################

Follow the steps in `"How to configure push updates" section of the openwisp-controller documentation 

> <https://github.com/openwisp/openwisp-controller#how-to-configure-push-updates>`_ to allow SSH access to you device from OpenWISP.

Author: Sure :+1:

quickly to events happening in the network and informs the user promptly
if there's anything that is not working as intended.

Iperf
~~~~~

This check provides network performance measurements such as maximum achievable bandwidth, 
jitter, datagram loss etc of the device using `iperf3 utility <https://iperf.fr/>`_. 

This check is **disabled by default**. You can enable auto creation of this check by setting the
`OPENWISP_MONITORING_AUTO_IPERF <#OPENWISP_MONITORING_AUTO_IPERF>`_ to ``True``.

It also supports tuning of various parameters.

> You can also change the parameters used for iperf checks (e.g. timing, port, username, password, rsa_publc_key etc) 

Reviewer: please break lines and maintain line length consistent with the rest of the document

This setting allows you to choose whether `config_applied <#configuration-applied>`_ checks should be
created automatically for newly registered devices. It's enabled by default.

``OPENWISP_MONITORING_AUTO_IPERF``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+-------------+
| **type**:    | ``bool``    |
+--------------+-------------+
| **default**: | ``False``   |
+--------------+-------------+

> Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices need to have SSH access from OpenWISP & must installed
> `iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ on devices in order for this check to work.

Reviewer 2: ```suggestion
Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices must have ``SSH`` `access credential <link to access credentials in controller module>`_  & 
`iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ must be installed on the devices in order for this check to work.
```

This setting allows you to choose whether `config_applied <#configuration-applied>`_ checks should be
created automatically for newly registered devices. It's enabled by default.

``OPENWISP_MONITORING_AUTO_IPERF``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+-------------+
| **type**:    | ``bool``    |
+--------------+-------------+
| **default**: | ``False``   |
+--------------+-------------+

> Whether `Iperf <#iperf-1>`_ checks are created automatically for devices. The devices must have ``SSH`` `access credential <https://github.com/openwisp/openwisp-controller#save-ssh-private-key-in-openwisp-access-credentials>`_  & 

Reviewer 2: I don't think we need to point out here that the device needs a working SSH connection here. 

Something like this would work fine 

```
This setting allows you to choose whether `iperf <#iperf-1>`_ checks should be
created automatically for newly registered devices. It's disabled by default.
```

`iperf3 openwrt package <https://openwrt.org/packages/pkgdata/iperf3>`_ must be installed on the devices in order for this check to work.

``OPENWISP_MONITORING_IPERF_CHECK_CONFIG``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+-------------+
| **type**:    | ``dict``    |
+--------------+-------------+
| **default**: | ``{}``      |
+--------------+-------------+

This setting allows to override the default iperf check configuration defined in
``openwisp_monitoring.check.classes.iperf.DEFAULT_IPERF_CHECK_CONFIG``.

For example, if you want to change only the **port number** of
``iperf`` check you can use:

.. code-block:: python

    OPENWISP_MONITORING_IPERF_CHECK_CONFIG = {
        'port': {
            'default': 6209,
        },
    }

> Similar to `OPENWISP_MONITORING_PING_CHECK_CONFIG <#OPENWISP_MONITORING_PING_CHECK_CONFIG>`_ , 

Reviewer 2: Let's remove reference to the configuration of Ping check. We will make users read the section even though we are explaining everything here.

This setting allows you to choose whether `config_applied <#configuration-applied>`_ checks should be
created automatically for newly registered devices. It's enabled by default.

``OPENWISP_MONITORING_AUTO_IPERF``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+-------------+
| **type**:    | ``bool``    |
+--------------+-------------+
| **default**: | ``False``   |
+--------------+-------------+

This setting allows you to choose whether `iperf <#iperf-1>`_ checks should be
created automatically for newly registered devices. It's disabled by default.

``OPENWISP_MONITORING_IPERF_CHECK_RSA_KEY_PATH``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

+--------------+-------------------------------+
| **type**:    | ``str``                       |
+--------------+-------------------------------+
| **default**: | ``/tmp/iperf-rsa-public.pem`` |
+--------------+-------------------------------+

This setting allows you to choose RSA public key path (relative to ``/root``)

> for `iperf check running with authentication <#configure-iperf-check-for-authentication>`_.

Reviewer: I don't understand, isn't this something that depends on a specific server? Why do we have a global setting?

Author: Yes it should be defined in [`OPENWISP_IPERF_CHECK_CONFIG`](https://github.com/openwisp/openwisp-monitoring/tree/issue-385/iperf-check#openwisp_monitoring_iperf_check_config)

defined in the ``alert_settings`` key.
You can always override them and define your own custom values via the *admin*.

You can also use ``alert_on_related_field`` key in metric configuration
which allows ``AlertSettings`` to use ``related_field`` as value to check ``threshold`` instead
of default ``field_name`` key. A real world example of this can be seen in 

> `Iperf metric configuration <https://github.com/openwisp/openwisp-monitoring/blob/issue-385/iperf-check/openwisp_monitoring/monitoring/configuration.py#L546-L700>`_,

Author: `Todo `: Change this link after PR merge

defined in the ``alert_settings`` key.
You can always override them and define your own custom values via the *admin*.

You can also use ``alert_on_related_field`` key in metric configuration
which allows ``AlertSettings`` to use ``related_field`` as value to check ``threshold`` instead
of default ``field_name`` key. A real world example of this can be seen in 
`Iperf metric configuration 

> <https://github.com/openwisp/openwisp-monitoring/blob/issue-385/iperf-check/openwisp_monitoring/monitoring/configuration.py#L546-L700>`_,

Reviewer: definitely we cannot link files which are published in branches we are going to delete, please double check the rest of the document for similar problems, we can point to master here, even thought it won't work until published, but that's ok.

``run_checks``
~~~~~~~~~~~~~~

> This command will execute all the `available checks <#available-checks>`_ for all the devices. 

Reviewer: this change is unnecessary, please revert it
